### ISSUE TITLE
Add feature or document how to reset Apollo store

### ISSUE THREAD

**Issue by stubailo** [apollostack/apollo-client#158]
For example, when the user logs out we might want to clear our some or all of the data on the client.

**Comment by abhiaiyer91** [comment]
There's way to achieve this today in Redux. We can compose a HO Root Reducer like

```js
function superMegaRootReducer(state, action) {
  switch (action.type) {
   case 'USER_LOGOUT':
      if (action.type === 'USER_LOGOUT') {
        state = undefined
     }
     return appReducer(state, action);
  }
};
```

What would be cool is next having a reset middleware, so when i want to do these resets, i can just hit an generic reset action type and handle some function i want to be called in the process.
https://gist.github.com/abhiaiyer91/a8a7ca18a8f9b11faa02c899165d942b

**Comment by stubailo** [comment]
Yep, just need to add some tests and make sure the client behaves correctly when all of the data disappears!

**Comment by abhiaiyer91** [comment]
Im thinking we should do the following

1. ship a Higher Order Reducer so people can reset state of any reducer they want.
2. Build in reset capability to `apolloReducer`
3. write tests
4. write docs

what do you think?

**Comment by stubailo** [comment]
What's the higher order reducer for?

**Comment by abhiaiyer91** [comment]
The higher order reducer is for people who want to reset their normal redux state based on the same state clearing the apolloReducer will have. Thoughts?

So essentially whatever we clear apolloReducer with i can wrap a redux reducer

`someReducer: composeResetReducer(someReducer, {})`

where composeResetReducer takes a reducer and some initial state

```
function composeResetReducer(reducer: Function, initialState: any): Function {
  return function composedReducer(state: any, action: Object): any {
    if (action.type === "RESET_STATE") {
      return initialState;
    }
    return reducer(state, action);
  };
}
```

**Comment by stubailo** [comment]
Hmm, I'd rather skip that for now - it doesn't seem relevant to Apollo IMO. I'd rather just document "Apollo uses the `RESET_STATE` action to reset, feel free to use that in your own reducer" or something like that.

I'd rather not get into the business of writing higher-order reducers for people, that sounds like a Redux thing.

But if you want to write a separate package for that, it would be fine to include that in the docs as a code example!

**Comment by abhiaiyer91** [comment]
Okay Step 1 function to use. https://github.com/abhiaiyer91/redux-reset

**Comment by timbotnik** [comment]
This is a real security issue for users on shared computers, is there currently a mechanism for clearing the state?

**Comment by helfer** [comment]
I think the current workaround is to explicitly pass the store to apollo on construction, and then remove the reference to that store. I think setting the store content to null wouldn't be enough, because you also want to get rid of history. But then again, I don't know much about Redux, so I'll defer to the experts.

**Comment by WilliamHolmes** [comment]
FYI:
Your global REDUX state can be reset using...
`type: RESET`

**Comment by timbotnik** [comment]
So I tried out a store reset using the mechanism described [here](http://stackoverflow.com/a/35641992), but I found the QueryManager threw an exception whenever I dispatched my RESET action.  

The internal state for query listeners is actually the issue.  If there are outstanding watched queries, it fails inside this function https://github.com/apollostack/apollo-client/blob/master/src/QueryManager.ts#L468-L477, since `queries` is `{}`, and then you end up here https://github.com/apollostack/apollo-client/blob/master/src/QueryManager.ts#L200 and explode when dereferencing `queryStoreValue`.
 
And just to clarify - there seems to be no issue if you make sure that all the watched queries are stopped BEFORE resetting the store (although I am not sure there isn't some other state that might need cleaning up as well).

From my perspective i think a `reset` API makes sense, since
1. Otherwise I’d have to learn about how to make a reset reducer myself and this is not exactly straightforward.  Actually just to add a reset reducer you have to create your own redux store for Apollo, and you lose some nice things like the crash reporter and the hooks for redux dev tools.
2. We have in-memory state that also needs to be cleaned up before we can safely clear the store.
3. We might not always use redux as a store?

**Comment by Poincare** [comment]
Should now be fixed w/ PR #314.

**Comment by rogchap** [comment]
Not sure if this is a complete example on how to reset the store (especially for a logout scenario). The `resetStore` func is hidden inside `QueryManager`; and there is no (clear) way to stop queries that are `in flight`.
